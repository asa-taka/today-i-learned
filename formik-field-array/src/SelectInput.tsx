import React from 'react'
import Downshift from 'downshift'

interface Props {

}

interface Item {
  name: string,
  age: number,
}

const items: Item[] = [
  { name: 'alice', age: 12 },
  { name: 'bob', age: 23 },
  { name: 'carol', age: 31 },
]

export default function SelectInput() {
  const onChange = (item: Item) => {
    console.log(item)
  }
  const itemToString = (item: Item) => item.name
  return (
    <Downshift>
      {({
        getInputProps,
        getItemProps,
        getLabelProps,
        getMenuProps,
        isOpen,
        inputValue,
        highlightedIndex,
        selectedItem,
      }) => (
        <div>
          <label {...getLabelProps()}>Enter a fruit</label>
          <input {...getInputProps()} />
          <ul {...getInputProps()}>
            {isOpen && items.map((item, index) => (
              <li {...getItemProps({
                key: item.name,
                index,
                item,
              })} />
            ))}
          </ul>
        </div>
      )}
    </Downshift>
  )
}
