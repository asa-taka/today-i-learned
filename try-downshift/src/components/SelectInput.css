.SelectInput-menuItem[aria-selected="true"] {
  background-color: #666;
}

.SelectInput-menu {
  border: 1px solid #999;
  padding: 1rem;
}
