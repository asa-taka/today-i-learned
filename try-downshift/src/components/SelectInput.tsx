import React, { useState, useEffect } from 'react'
import Downshift, { ControllerStateAndHelpers as Controller } from 'downshift'
import { User, IdType, queryUsers } from '../api'
import './SelectInput.css'

interface Props {
  onSelect: (id: IdType) => void
  exclude?: IdType[]
}

export default function SelectInput(props: Props) {
  const { onSelect, exclude } = props
  const [items, setItems] = useState<User[]>([])

  useEffect(() => {
    queryUsers('').then(setItems)
  }, [])

  const handleSelect = (user: User, ctrl: Controller<User>) => {
    if (onSelect && user) onSelect(user.id)
    ctrl.clearSelection()
    ctrl.openMenu()
  }

  const handleInputValueChange = (input: string) => {
    queryUsers(input).then(setItems)
  }

  const itemsToUse = exclude ? items.filter(e => !exclude.includes(e.id)) : items

  return (
    <Downshift
      onChange={handleSelect}
      onInputValueChange={handleInputValueChange}
      itemToString={e => e.name}
    >
      {({
        getInputProps,
        getItemProps,
        getMenuProps,
        isOpen,
        openMenu,
      }) => (
        <div className="SelectInput-root">
          <input className="SelectInput-input" {...getInputProps()} onFocus={() => openMenu()} />
          {isOpen && (
            <div className="SelectInput-menu" {...getMenuProps()}>
              {itemsToUse.map((item, index) => (
                <div className="SelectInput-menuItem" key={item.id} {...getItemProps({ item, index })}>
                  {item.name}
                </div>
              ))}
            </div>
          )}
          <button className="SelectInput-submit" type="submit">submit</button>
        </div>
      )}
    </Downshift>
  )
}
