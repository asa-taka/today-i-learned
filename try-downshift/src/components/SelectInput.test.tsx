import React from 'react'
import { render, getByText, fireEvent, wait, waitForElement, waitForElementToBeRemoved } from '@testing-library/react'
import SelectInput from './SelectInput'

function setup() {
  const props = {
    onSelect: jest.fn()
  }
  const helpers = render(<SelectInput {...props} />)
  const { container } = helpers
  const getInput = () => container.querySelector('input')
  const getMenu = () => container.querySelector('[role="listbox"]') as HTMLElement
  const getMenuItem = (text: string) => getByText(getMenu(), text)
  return {
    ...helpers,
    props,
    getInput,
    getMenu,
    getMenuItem,
  }
}

describe('SelectInput', () => {
  it('renders without crush', async () => {
    const { container } = setup()
    expect(container).toMatchSnapshot()
  })

  it('shows menu on focus', async () => {
    const { container, getInput, getMenu } = setup()
    getInput().focus()
    await wait(() => {
      expect(getMenu()).toBeDefined()
    })
    expect(container).toMatchSnapshot()
  })

  it('queries by input value', async () => {
    const { container, getInput, getMenuItem } = setup()
    getInput().focus()
    fireEvent.change(getInput(), { target: { value: 'alice' }})
    await waitForElement(() => getMenuItem('alice'))
    expect(container).toMatchSnapshot()
  })

  it('fires `onSelect` once for each user-lick', async () => {
    const { container, props, getInput, getMenu, getMenuItem } = setup()
    getInput().focus()
    await waitForElement(() => getMenuItem('alice'))
    fireEvent.click(getMenuItem('alice'))
    expect(props.onSelect.mock.calls.length).toBe(1)
    expect(container).toMatchSnapshot()
  })
})
