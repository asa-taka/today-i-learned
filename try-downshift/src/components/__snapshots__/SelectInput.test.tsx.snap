// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectInput fires \`onSelect\` once for each user-lick 1`] = `
<div>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-3-label"
    aria-owns="downshift-3-menu"
    class="SelectInput-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-3-menu"
      aria-labelledby="downshift-3-label"
      autocomplete="off"
      class="SelectInput-input"
      id="downshift-3-input"
      value=""
    />
    <div
      aria-labelledby="downshift-3-label"
      class="SelectInput-menu"
      id="downshift-3-menu"
      role="listbox"
    >
      <div
        aria-selected="false"
        class="SelectInput-menuItem"
        id="downshift-3-item-0"
        role="option"
      >
        alice
      </div>
      <div
        aria-selected="false"
        class="SelectInput-menuItem"
        id="downshift-3-item-1"
        role="option"
      >
        bobby
      </div>
      <div
        aria-selected="false"
        class="SelectInput-menuItem"
        id="downshift-3-item-2"
        role="option"
      >
        cancy
      </div>
      <div
        aria-selected="false"
        class="SelectInput-menuItem"
        id="downshift-3-item-3"
        role="option"
      >
        daddy
      </div>
      <div
        aria-selected="false"
        class="SelectInput-menuItem"
        id="downshift-3-item-4"
        role="option"
      >
        evily
      </div>
    </div>
    <button
      class="SelectInput-submit"
      type="submit"
    >
      submit
    </button>
  </div>
</div>
`;

exports[`SelectInput queries by input value 1`] = `
<div>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-2-label"
    aria-owns="downshift-2-menu"
    class="SelectInput-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-2-menu"
      aria-labelledby="downshift-2-label"
      autocomplete="off"
      class="SelectInput-input"
      id="downshift-2-input"
      value="alice"
    />
    <div
      aria-labelledby="downshift-2-label"
      class="SelectInput-menu"
      id="downshift-2-menu"
      role="listbox"
    >
      <div
        aria-selected="false"
        class="SelectInput-menuItem"
        id="downshift-2-item-0"
        role="option"
      >
        alice
      </div>
    </div>
    <button
      class="SelectInput-submit"
      type="submit"
    >
      submit
    </button>
  </div>
</div>
`;

exports[`SelectInput renders without crush 1`] = `
<div>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="downshift-0-label"
    class="SelectInput-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-labelledby="downshift-0-label"
      autocomplete="off"
      class="SelectInput-input"
      id="downshift-0-input"
      value=""
    />
    <button
      class="SelectInput-submit"
      type="submit"
    >
      submit
    </button>
  </div>
</div>
`;

exports[`SelectInput shows menu on focus 1`] = `
<div>
  <div
    aria-expanded="true"
    aria-haspopup="listbox"
    aria-labelledby="downshift-1-label"
    aria-owns="downshift-1-menu"
    class="SelectInput-root"
    role="combobox"
  >
    <input
      aria-autocomplete="list"
      aria-controls="downshift-1-menu"
      aria-labelledby="downshift-1-label"
      autocomplete="off"
      class="SelectInput-input"
      id="downshift-1-input"
      value=""
    />
    <div
      aria-labelledby="downshift-1-label"
      class="SelectInput-menu"
      id="downshift-1-menu"
      role="listbox"
    />
    <button
      class="SelectInput-submit"
      type="submit"
    >
      submit
    </button>
  </div>
</div>
`;
